src/za/ac/sun/cs/green/service/simplify/ConstantPropogation.java

```diff
@@ -0,0 +1,129 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.SortedSet;
+import java.util.Stack;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.logging.Level;
+
+import za.ac.sun.cs.green.Instance;
+import za.ac.sun.cs.green.Green;
+import za.ac.sun.cs.green.expr.Expression;
+import za.ac.sun.cs.green.service.BasicService;
+import za.ac.sun.cs.green.util.Reporter;
+import za.ac.sun.cs.green.expr.Constant;
+import za.ac.sun.cs.green.expr.IntConstant;
+import za.ac.sun.cs.green.expr.IntVariable;
+import za.ac.sun.cs.green.expr.Operation;
+import za.ac.sun.cs.green.expr.Variable;
+import za.ac.sun.cs.green.expr.Visitor;
+import za.ac.sun.cs.green.expr.VisitorException;
+
+public class ConstantPropogation extends BasicService {
+
+	private int invocations = 0;
+
+	public ConstantPropogation(Green solver) {
+		super(solver);
+	}
+
+	@Override
+	public Set<Instance> processRequest(Instance instance) {
+		@SuppressWarnings("unchecked")
+		Set<Instance> result = (Set<Instance>) instance.getData(getClass());
+		if (result == null) {
+			final Map<Variable, Variable> map = new HashMap<Variable, Variable>();
+			final Expression e = canonize(instance.getFullExpression(), map);
+			final Instance i = new Instance(getSolver(), instance.getSource(), null, e);
+			result = Collections.singleton(i);
+			instance.setData(getClass(), result);
+		}
+		return result;
+	}
+
+	@Override
+	public void report(Reporter reporter) {
+		reporter.report(getClass().getSimpleName(), "invocations = " + invocations);
+	}
+
+	public Expression canonize(Expression expression, Map<Variable, Variable> map) {
+		try {
+			log.log(Level.FINEST, "Before Canonization: " + expression);
+			invocations++;
+			ConstantPropVisitor constantPropVisitor = new ConstantPropVisitor();
+			expression.accept(constantPropVisitor);
+			expression = constantPropVisitor.getExpression();
+			log.log(Level.FINEST, "After Canonization: " + expression);
+			return expression;
+		} catch (VisitorException x) {
+			log.log(Level.SEVERE, "encountered an exception -- this should not be happening!", x);
+		}
+		return null;
+	}
+
+	private static class ConstantPropVisitor extends Visitor {
+
+		private Stack<Expression> stack;
+		private HashMap<String, Integer> map;
+
+		public ConstantPropVisitor() {
+			stack = new Stack<Expression>();
+			map = new HashMap<String, Integer>();
+		}
+
+		public Expression getExpression() {
+			// Print HashMap
+			// System.out.println(Arrays.asList(map));
+			return stack.pop();
+		}
+
+		@Override
+		public void postVisit(IntConstant constant) {
+			stack.push(constant);
+		}
+
+		@Override
+		public void postVisit(IntVariable variable) {
+			stack.push(variable);
+		}
+
+		@Override
+		public void postVisit(Operation operation) throws VisitorException {
+			Expression r = stack.pop();
+			Expression l = stack.pop();
+
+			if (operation.getOperator() == Operation.Operator.EQ
```

Comment:
You could probably simplify some of your conditionals a bit. It'll add to your line count, but you are gaining with respect to readability.

I am also fully aware that the rest of the Green codebase tends not to follow this mantra. But as per usual, be the change you want to see in the world etc.

Reply:
The actual complexity starts on line 102-103, but the little preview tool shows only what it sees fit to show :imp: